src/runtime/allocator.ts

```diff
@@ -0,0 +1,291 @@
+    }));
+  }
+
+  async assignStorageKeys(arcId: ArcId, plan: Recipe, idGenerator?: IdGenerator): Promise<Recipe> {
+    // TODO(b/182410550): All internal caller(s) should pass non normalized recipe.
+    // Remove this check, once the method is private, and don't return recipe.
+    if (plan.isNormalized()) {
+      plan = plan.clone();
+    }
+    // Assign storage keys for all `create` & `copy` stores.
+    for (const handle of plan.handles) {
+      if (handle.immediateValue) continue;
+      if (['copy', 'create'].includes(handle.fate)) {
+      let type = handle.type;
+        if (handle.fate === 'create') {
+          assert(type.maybeResolve(), `Can't assign resolved type to ${type}`);
+        }
+
+        type = type.resolvedType();
+        assert(type.isResolved(), `Can't create handle for unresolved type ${type}`);
+        handle.id = handle.fate === 'create' && !!handle.id
+          ? handle.id
+          : (idGenerator || this.arcStateById.get(arcId).idGenerator).newChildId(arcId, '').toString();
+        handle.fate = 'use';
+        handle.storageKey = await this.runtime.getCapabilitiesResolver(arcId)
+          .createStorageKey(handle.capabilities || Capabilities.create(), type, handle.id);
+      }
+    }
+    return this.resolveRecipe(arcId, plan);
+  }
+
+  // Returns the resolved recipe (the original recipe might have changed),
+  // or throws an exception, if the recipe cannot be resolved.
+  // Note this method is overriden by the subclass, and calls RecipeResolver, if needed.
+  // The APIs might change, when Planning is incorporated in Allocator.
+  protected async resolveRecipe(arcId: ArcId, recipe: Recipe): Promise<Recipe> {
+    assert(this.tryResolveRecipe(arcId, recipe), `Unresolved recipe: ${recipe.toString({showUnresolved: true})}`);
+    return recipe;
+  }
+
+  // Normalizes, if needed, and tries to resolve type variable handle types.
+  // Returns true, if the recipe is resolved.
+  protected tryResolveRecipe(arcId: ArcId, recipe: Recipe): boolean {
+    assert(this.normalize(recipe));
+    if (!recipe.isResolved()) {
+      for (const handle of recipe.handles) {
+        // The call to `normalize` above un-resolves typevar handle types.
+        assert(handle.type.maybeResolve());
+      }
+    }
+    return recipe.isResolved();
+  }
+
+  private normalize(recipe: Recipe): boolean {
+    if (recipe.isNormalized()) {
+      return true;
+    }
+    const errors = new Map();
+    if (recipe.normalize({errors})) {
+      return true;
+    }
+    console.warn('failed to normalize:\n', errors, recipe.toString());
+    return false;
+  }
+
+  public stopArc(arcId: ArcId) {
+    assert(this.arcStateById.get(arcId));
+    for (const partition of this.arcStateById.get(arcId).partitions) {
+      const host = this.hostById[partition.arcHostId];
+      assert(host);
+      host.stop(arcId);
+    }
+    this.arcStateById.delete(arcId);
+  }
+
+  async deserialize(options: DeserializeArcOptions): Promise<ArcId> {
```

Comment:
(OK for this CL, for future discussion): would be worth figuring out how this works in a bit more detail. I think the trend in Kotlin at the moment is to separate out (de)serialization - at least partially.

Reply:
I thought de/serialization was simply disabled in kotlin, but was part of the Allocator, but i must be wrong. 
Anyway, yes, if that's fine for now, let's leave it here.